tests/cupy_tests/core_tests/fusion_tests/test_function.py

```diff
@@ -0,0 +1,553 @@
+            scalar1_copy = scalar1
+            scalar1 += scalar2
+            return array + scalar1 + scalar1_copy
+
+        return func
+
+    @testing.for_all_dtypes_combination(names=('dtype1', 'dtype2'))
+    @fusion_utils.check_fusion(generate_inputs_name='numpy_scalar_param')
+    def test_scalar_inplace_update_with_array(self, xp, dtype1, dtype2):
+        def func(array, scalar):
+            scalar += array
+            return scalar
+
+        return func
+
+
+@testing.gpu
+class TestFusionBroadcast(unittest.TestCase):
+
+    def generate_inputs(self, xp, dtype):
+        x = testing.shaped_random((3, 4), xp, dtype, scale=10, seed=0)
+        y = testing.shaped_random((4,), xp, dtype, scale=10, seed=1)
+        return (x, y), {}
+
+    @testing.for_all_dtypes()
+    @fusion_utils.check_fusion()
+    def test_broadcast(self, xp, dtype):
+        def func(x, y):
+            x += y
+            return x
+
+        return func
+
+    @testing.for_all_dtypes()
+    @fusion_utils.check_fusion(accept_error=ValueError)
+    def test_broadcast_error(self, xp, dtype):
+        def func(x, y):
+            y += x
+            return y
+
+        return func
+
+
+@testing.gpu
+class TestFusionNoneParams(unittest.TestCase):
+
+    @testing.for_all_dtypes()
+    @testing.numpy_cupy_array_equal()
+    def test_python_none_parameter(self, xp, dtype):
+        @cupy.fuse()
+        def f(x, y, z):
+            if y is None:
+                return x * z
+            return x + y + z
+
+        x = testing.shaped_arange((10,), xp, dtype)
+        y = testing.shaped_arange((10,), xp, dtype)
+        z = testing.shaped_arange((10,), xp, dtype)
+        return f(x, None, z) + f(x, y, z)
+
+
+@testing.gpu
+class TestFusionReturnsConstantValue(FusionTestBase):
+
+    @testing.for_all_dtypes()
+    @fusion_utils.check_fusion(generate_inputs_args=(1,))
+    def test_pass(self, xp, dtype):
+        def func(x):
+            pass
+
+        return func
+
+    @testing.for_all_dtypes(no_bool=True)
+    @fusion_utils.check_fusion(generate_inputs_args=(1,))
+    def test_no_return_value(self, xp, dtype):
+        def func(x):
+            x += 1
+
+        return func
+
+
+class TestFusionComposition(unittest.TestCase):
+
+    @testing.for_all_dtypes(no_bool=True)
+    @testing.numpy_cupy_array_equal()
+    def test_composition(self, xp, dtype):
+        @cupy.fuse()
+        def f(x, y):
+            return x - y * 2, x + y
+
+        @cupy.fuse()
+        def g(x, y, z):
+            a, b = f(x + z, z - x * 3)
+            c, d = f(x - y, y - z)
+            return a + b * c - d
+
+        @cupy.fuse()
+        def h(x, y):
+            a, b = f(x + y * 2, y * 3)
+            return a - b * g(x - 2, x - 3, -y)
+
+        x = testing.shaped_arange((3, 3), xp, dtype)
+        y = testing.shaped_arange((3, 3), xp, dtype)
+        return h(x, y)
+
+
+class TestFusionCompile(unittest.TestCase):
+
+    # TODO(asi1024): Support it
```

Comment:
Be more specific about what to support.
What's `_compile_from_dtypes`?

Reply:
Oh this is not directly related to this PR.

Reply:
Removed this commented test, because it was a test for a private method.